#include "widget.h"
#include "ui_widget.h"
#define VERSION "V1.0"

Widget::Widget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::Widget)
{
    setWindowIcon(QIcon(":/logo.ico"));
    ui->setupUi(this);
    this->setWindowTitle("English Unfamiliar Words Extractorï¼ˆEUWEï¼‰\t"+ QString(VERSION));
    this->setFixedSize(this->width(),this->height());
    wdb = new WordsDatabase(this);

    // è¿æ¥ä¿¡å·æ§½
    connect(ui->openDb,&QPushButton::clicked,[=](){
        wdb->show();
    });

    connect(ui->help,&QPushButton::clicked,[=](){
        QMessageBox::about(nullptr,"å¸®åŠ©","è”ç³»Bcaid\nQQ:75011847");
    });

    connect(ui->about,&QPushButton::clicked,[=](){
        QMessageBox::about(nullptr,"å…³äº","è½¯ä»¶ç‰ˆæœ¬:" + QString(VERSION)+"\n ç”±Bcaidç¼–å†™");
    });

    connect(ui->leftextract,&QPushButton::clicked,[=](){
        analyze(ui->textEdit->toPlainText());
    });

    // ===== æ–°å¢æŒ‰é’®åŠŸèƒ½ =====

    // æ¸…ç©ºæŒ‰é’®
    connect(ui->clearBtn, &QPushButton::clicked, [=]() {
        ui->textEdit->clear();
        ui->statusLabel->setText("âœ¨ æ–‡æœ¬å·²æ¸…ç©º");
        updateStats(); // æ›´æ–°ç»Ÿè®¡ä¿¡æ¯
    });

    // å¤åˆ¶æŒ‰é’®
    connect(ui->copyBtn, &QPushButton::clicked, [=]() {
        QString text = ui->textEdit->toPlainText();
        if (text.isEmpty()) {
            ui->statusLabel->setText("âš ï¸ æ²¡æœ‰æ–‡æœ¬å¯å¤åˆ¶");
            return;
        }

        QClipboard *clipboard = QApplication::clipboard();
        clipboard->setText(text);
        ui->statusLabel->setText("âœ… å·²å¤åˆ¶åˆ°å‰ªè´´æ¿");
    });

    // æ–‡æœ¬å†…å®¹å˜åŒ–æ—¶æ›´æ–°ç»Ÿè®¡
    connect(ui->textEdit, &QTextEdit::textChanged, [=]() {
        updateStats();
    });

    // æ–‡ä»¶å¯¼å…¥æŒ‰é’® - æ”¯æŒ txt, pdf, doc/docx
    connect(ui->input, &QPushButton::clicked, [=]() {
        // å¼¹å‡ºæ–‡ä»¶é€‰æ‹©å¯¹è¯æ¡†ï¼Œè¿‡æ»¤å™¨ï¼šæ‰€æœ‰æ”¯æŒçš„æ–‡ä»¶
        QString filter = "æ‰€æœ‰æ–‡ä»¶ (*.*);;æ–‡æœ¬æ–‡æ¡£ (*.txt);;PDF æ–‡ä»¶ (*.pdf);;Word æ–‡æ¡£ (*.doc *.docx)";
        QString filePath = QFileDialog::getOpenFileName(this, "é€‰æ‹©æ–‡ä»¶", QString(), filter);
        if (filePath.isEmpty()) {
            return; // ç”¨æˆ·å–æ¶ˆ
        }

        // æ›´æ–°çŠ¶æ€
        ui->statusLabel->setText("â³ æ­£åœ¨è¯»å–æ–‡ä»¶...");
        QApplication::processEvents(); // è®©ç•Œé¢åŠæ—¶åˆ·æ–°

        QString content;
        QFileInfo fileInfo(filePath);
        QString suffix = fileInfo.suffix().toLower();

        // æ ¹æ®åç¼€é€‰æ‹©è¯»å–å‡½æ•°
        if (suffix == "txt") {
            content = readTxt(filePath);
        } else if (suffix == "pdf") {
            content = readPdf(filePath);
        } else if (suffix == "doc" || suffix == "docx") {
            content = readWordDocument(filePath);
        } else {
            // å°è¯•ä½œä¸ºçº¯æ–‡æœ¬æ‰“å¼€ï¼ˆç”¨æˆ·å¯èƒ½é€‰äº†æ‰€æœ‰æ–‡ä»¶ï¼‰
            content = readTxt(filePath);
        }

        // æ£€æŸ¥æ˜¯å¦è¯»å–æˆåŠŸ
        if (content.isEmpty()) {
            QMessageBox::warning(this, "è¯»å–å¤±è´¥", "æ— æ³•è¯»å–æ–‡ä»¶å†…å®¹ï¼Œè¯·æ£€æŸ¥æ–‡ä»¶æ ¼å¼æˆ–æƒé™ã€‚");
            ui->statusLabel->setText("âŒ è¯»å–å¤±è´¥");
        } else {
            ui->textEdit->setPlainText(content);
            ui->statusLabel->setText("âœ… æ–‡ä»¶å¯¼å…¥æˆåŠŸï¼š" + fileInfo.fileName());
            updateStats(); // æ›´æ–°å­—ç¬¦/å•è¯ç»Ÿè®¡
        }
    });

    // OCRæŒ‰é’®åŠŸèƒ½
    connect(ui->ocr, &QPushButton::clicked, [=]() {
        // 1. å¼¹å‡ºæ–‡ä»¶å¯¹è¯æ¡†é€‰æ‹©å›¾ç‰‡
        QString filter = "å›¾ç‰‡æ–‡ä»¶ (*.png *.jpg *.jpeg *.bmp *.tiff *.gif);;æ‰€æœ‰æ–‡ä»¶ (*.*)";
        QString imagePath = QFileDialog::getOpenFileName(this, "é€‰æ‹©è¦è¯†åˆ«çš„å›¾ç‰‡", QString(), filter);
        if (imagePath.isEmpty()) {
            return; // ç”¨æˆ·å–æ¶ˆ
        }

        // æ›´æ–°çŠ¶æ€
        ui->statusLabel->setText("â³ æ­£åœ¨è¯†åˆ«å›¾ç‰‡...");
        QApplication::processEvents();

        // 2. åˆ›å»º QProcess å¯¹è±¡
        QProcess *ocrProcess = new QProcess(this);

        // 3. è¿æ¥ä¿¡å·ï¼šè¯»å–è¾“å‡º
        connect(ocrProcess, &QProcess::readyReadStandardOutput, this, [=]() {
            QByteArray output = ocrProcess->readAllStandardOutput();
            QString result = QString::fromUtf8(output).trimmed();
            if (!result.isEmpty()) {
                // å°†è¯†åˆ«ç»“æœè¿½åŠ åˆ°æ–‡æœ¬ç¼–è¾‘åŒºï¼ˆä¹Ÿå¯é€‰æ‹©æ›¿æ¢ï¼Œè¿™é‡Œé‡‡ç”¨è¿½åŠ å¹¶æ¢è¡Œï¼‰
                QString currentText = ui->textEdit->toPlainText();
                if (!currentText.isEmpty() && !currentText.endsWith('\n')) {
                    ui->textEdit->append(""); // æ¢è¡Œ
                }
                ui->textEdit->append("ã€OCRè¯†åˆ«ç»“æœã€‘");
                ui->textEdit->append(result);
                ui->textEdit->append(""); // ç©ºè¡Œåˆ†éš”
            }
        });

        // 4. è¿æ¥ä¿¡å·ï¼šè¯»å–é”™è¯¯è¾“å‡º
        connect(ocrProcess, &QProcess::readyReadStandardError, this, [=]() {
            QByteArray err = ocrProcess->readAllStandardError();
            qDebug() << "OCR stderr:" << QString::fromUtf8(err);
        });

        // 5. è¿›ç¨‹ç»“æŸå¤„ç†
        connect(ocrProcess, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
                this, [=](int exitCode, QProcess::ExitStatus exitStatus) {
                    if (exitStatus == QProcess::CrashExit) {
                        ui->statusLabel->setText("âŒ OCRè¿›ç¨‹å´©æºƒ");
                        qDebug() << "OCR è¿›ç¨‹å´©æºƒ";
                    } else if (exitCode != 0) {
                        ui->statusLabel->setText("âŒ OCRè¯†åˆ«å¤±è´¥ï¼Œé€€å‡ºç : " + QString::number(exitCode));
                        qDebug() << "OCR è¿›ç¨‹é”™è¯¯ï¼Œé€€å‡ºç :" << exitCode;
                    } else {
                        ui->statusLabel->setText("âœ… OCRè¯†åˆ«å®Œæˆ");
                        qDebug() << "OCR å®Œæˆï¼Œé€€å‡ºç :" << exitCode;
                    }
                    ocrProcess->deleteLater(); // æ¸…ç†è¿›ç¨‹å¯¹è±¡
                });

        // 6. è¿›ç¨‹å¯åŠ¨å¤±è´¥å¤„ç†
        connect(ocrProcess, &QProcess::errorOccurred, this, [=](QProcess::ProcessError error) {
            QString errorMsg;
            switch (error) {
            case QProcess::FailedToStart:
                errorMsg = "å¯åŠ¨å¤±è´¥ï¼Œè¯·ç¡®è®¤ EUWEOCRT.exe æ˜¯å¦å­˜åœ¨ä¸”å¯æ‰§è¡Œ";
                break;
            case QProcess::Crashed:
                errorMsg = "è¿›ç¨‹å´©æºƒ";
                break;
            case QProcess::Timedout:
                errorMsg = "ç­‰å¾…è¶…æ—¶";
                break;
            case QProcess::WriteError:
                errorMsg = "å†™å…¥é”™è¯¯";
                break;
            case QProcess::ReadError:
                errorMsg = "è¯»å–é”™è¯¯";
                break;
            default:
                errorMsg = "æœªçŸ¥é”™è¯¯";
            }
            ui->statusLabel->setText("âŒ OCRè¿›ç¨‹é”™è¯¯: " + errorMsg);
            qDebug() << "å¯åŠ¨ OCR è¿›ç¨‹å¤±è´¥:" << error;
            ocrProcess->deleteLater();
        });

        // 7. è®¾ç½®ç¨‹åºè·¯å¾„å’Œå‚æ•°
        // å‡è®¾ EUWEOCRT.exe ä¸ä¸»ç¨‹åºåœ¨åŒä¸€ç›®å½•ä¸‹
        QString program = QCoreApplication::applicationDirPath() + "/EUWEOCRT.exe";
        // å¦‚æœå¸Œæœ›ä½¿ç”¨ç›¸å¯¹è·¯å¾„ï¼ˆç¨‹åºè¿è¡Œç›®å½•ï¼‰ï¼Œå¯ç›´æ¥ç”¨ "EUWEOCRT.exe"
        // QString program = "EUWEOCRT.exe";

        QStringList arguments;
        arguments << imagePath;  // ä¼ é€’å›¾ç‰‡è·¯å¾„

        // å¯é€‰ï¼šæŒ‡å®š tessdata è·¯å¾„å’Œè¯­è¨€ï¼ˆæ ¹æ®ä½ çš„ OCR åŠ©æ‰‹æ”¯æŒçš„å‚æ•°ï¼‰
        // arguments << "-t" << QCoreApplication::applicationDirPath() + "/tessdata";
        // arguments << "-l" << "eng+chi_sim";

        // 8. å¯åŠ¨è¿›ç¨‹
        ocrProcess->start(program, arguments);
        if (!ocrProcess->waitForStarted(3000)) { // ç­‰å¾…æœ€å¤š3ç§’å¯åŠ¨
            ui->statusLabel->setText("âŒ OCRè¿›ç¨‹å¯åŠ¨è¶…æ—¶");
            ocrProcess->deleteLater();
        }
    });

    // åˆå§‹åŒ–çŠ¶æ€
    updateStats();
    ui->statusLabel->setText("âœ¨ å°±ç»ª - å¯ä»¥å¼€å§‹æå–æ–‡æœ¬");
}

Widget::~Widget()
{
    delete ui;
}


void Widget::analyze(QString str){
    EnglishExtracter ee(str);
    unordered_set<QString> pre_words = ee.get();

    vector<QString> notwords;
    vector<QString> yzwords;//å·²çŸ¥
    vector<Word> words;

    for(QString word : pre_words){

        QApplication::processEvents();//é˜²æ­¢é¡µé¢å¡æ­»

        QList<int> matchRows;
        QList<QString> match;
        QList<QString> match2;
        QList<QString> match3;

        if(wdb->findWithRowNumbers(word,matchRows,match)){
            yzwords.push_back(word);
            continue;
        }
        if(!wdb->find(word,matchRows,match,match2,match3)){
            notwords.push_back(word);
            continue;
        }
        words.push_back(Word(word,match2[0],match3[0]));
    }

    // è°ƒè¯•è¾“å‡º
    for(QString w : yzwords){
        qDebug() << "ç†Ÿè¯: " << w;
    }
    for(QString w : notwords){
        qDebug() << "æœªè¯†åˆ«: " << w;
    }
    for(Word w : words){
        qDebug() << "å·²è¯†åˆ«: " << w.word  << "\t"<< w.phonetic << "\t"<< w.translation;
    }

    // æ˜¾ç¤ºç»“æœçª—å£
    ResultWidget* resultWidget = new ResultWidget();
    resultWidget->setDatabase(wdb);  // è®¾ç½®æ•°æ®åº“æŒ‡é’ˆ
    resultWidget->setNotWords(notwords);
    resultWidget->setYzWords(yzwords);
    resultWidget->setWords(words);
    resultWidget->show();

    // è¿æ¥ç»§ç»­è¯†åˆ«ä¿¡å·
    connect(resultWidget, &ResultWidget::continueRequested, [=]() {
        // å¯ä»¥åœ¨è¿™é‡Œæ·»åŠ ç»§ç»­è¯†åˆ«çš„é€»è¾‘
        ui->textEdit->setFocus();
    });

    // æ›´æ–°çŠ¶æ€
    ui->statusLabel->setText(QString("âœ… åˆ†æå®Œæˆ - ç†Ÿè¯: %1, ç”Ÿè¯: %2, æœªè¯†åˆ«: %3")
                                 .arg(yzwords.size())
                                 .arg(words.size())
                                 .arg(notwords.size()));
}

// æ›´æ–°ç»Ÿè®¡ä¿¡æ¯
void Widget::updateStats()
{
    QString text = ui->textEdit->toPlainText();

    // å­—ç¬¦æ•°ï¼ˆä¸åŒ…æ‹¬ç©ºæ ¼å’Œæ¢è¡Œï¼‰
    int charCount = 0;
    for (QChar ch : text) {
        if (!ch.isSpace()) {
            charCount++;
        }
    }

    // å•è¯æ•°
    QStringList words = text.split(QRegularExpression("\\s+"), Qt::SkipEmptyParts);
    int wordCount = words.size();

    // è¡Œæ•°
    int lineCount = text.split('\n').size();
    if (text.isEmpty()) lineCount = 0;

    // æ›´æ–°UI
    ui->statsLabel->setText(QString("ğŸ“Š å­—ç¬¦æ•°: %1 | å•è¯æ•°: %2 | è¡Œæ•°: %3")
                                .arg(charCount)
                                .arg(wordCount)
                                .arg(lineCount));
}


QString Widget::readTxt(const QString &filePath) {
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qWarning() << "æ— æ³•æ‰“å¼€æ–‡ä»¶ï¼š" << filePath;
        return QString();
    }

    QTextStream stream(&file);
    // å¯ç”¨è‡ªåŠ¨æ£€æµ‹ç¼–ç ï¼ˆå¯ä»¥è¯†åˆ«å¸¦ BOM çš„ UTF-8/UTF-16 ç­‰ï¼‰
    stream.setAutoDetectUnicode(true);
    QString content = stream.readAll();
    file.close();

    return content;
}

QString Widget::readPdf(const QString &filePath) {
    QString allText;

    // 1. åŠ è½½ PDF æ–‡æ¡£
    QPdfDocument document;
    QPdfDocument::Error error = document.load(filePath);
    if (error != QPdfDocument::Error::None) {
        qWarning() << "æ— æ³•åŠ è½½ PDF æ–‡ä»¶ï¼š" << filePath << "ï¼Œé”™è¯¯ä»£ç ï¼š" << static_cast<int>(error);
        return allText;
    }

    int pageCount = document.pageCount();
    qDebug() << "PDF åŠ è½½æˆåŠŸï¼Œæ€»é¡µæ•°ï¼š" << pageCount;

    // 2. éå†æ¯ä¸€é¡µæå–æ–‡æœ¬
    for (int pageIndex = 0; pageIndex < pageCount; ++pageIndex) {
        // ä½¿ç”¨ getAllText è·å–æ•´é¡µæ–‡æœ¬
        QPdfSelection selection = document.getAllText(pageIndex);

        if (selection.isValid() && !selection.text().isEmpty()) {
            allText.append(selection.text());
        } else {
            qWarning() << "ç¬¬" << pageIndex + 1 << "é¡µæ²¡æœ‰æå–åˆ°æ–‡æœ¬";
        }

        QApplication::processEvents();//é˜²æ­¢é¡µé¢å¡æ­»
        // åœ¨é¡µé¢ä¹‹é—´æ·»åŠ æ¢è¡Œç¬¦ï¼ˆå¯é€‰ï¼‰
        if (pageIndex < pageCount - 1) {
            allText.append('\n');
        }
    }

    document.close();

    if (allText.isEmpty()) {
        qWarning() << "PDF æ–‡ä»¶æ²¡æœ‰æå–åˆ°ä»»ä½•æ–‡æœ¬";
    } else {
        qDebug() << "PDF æ–‡æœ¬æå–å®Œæˆï¼Œæ€»å­—ç¬¦æ•°ï¼š" << allText.length();
    }

    return allText;
}

// è¯»å–Wordæ–‡æ¡£æ‰€æœ‰æ–‡æœ¬çš„å‡½æ•°
QString Widget::readWordDocument(const QString &filePath) {
    QString allText;

    // 1. åˆ›å»ºWordåº”ç”¨ç¨‹åºå¯¹è±¡
    QAxObject *wordApp = new QAxObject("Word.Application", nullptr);
    if (!wordApp) {
        qDebug() << "Failed to initialize Word application.";
        return allText;
    }

    // å»ºè®®åœ¨è°ƒè¯•æ—¶å°†Visibleè®¾ä¸ºtrueï¼Œä»¥ä¾¿è§‚å¯ŸWordç¨‹åºçš„å¯åŠ¨
    // wordApp->setProperty("Visible", true);

    // 2. è·å–æ–‡æ¡£é›†åˆå¯¹è±¡
    QAxObject *documents = wordApp->querySubObject("Documents");

    // 3. æ‰“å¼€æŒ‡å®šè·¯å¾„çš„æ–‡æ¡£
    // æ³¨æ„ï¼šOpenæ–¹æ³•çš„å‚æ•°å–å†³äºä½ çš„éœ€æ±‚ï¼Œè¿™é‡Œç¬¬äºŒä¸ªå‚æ•°è¡¨ç¤ºæ˜¯å¦åªè¯»
    QAxObject *document = documents->querySubObject("Open(const QString&, bool)", filePath, true);
    if (!document) {
        qDebug() << "Failed to open document:" << filePath;
        wordApp->dynamicCall("Quit()");
        delete wordApp;
        return allText;
    }

    // 4. è·å–æ–‡æ¡£çš„æ‰€æœ‰æ–‡æœ¬
    // é€šè¿‡è·å–æ–‡æ¡£çš„Rangeå¯¹è±¡ï¼Œç„¶åè¯»å–å…¶Textå±æ€§ [citation:7]
    QAxObject *range = document->querySubObject("Range()");
    if (range) {
        allText = range->property("Text").toString();
        delete range;
    }

    // æˆ–è€…ï¼Œä½ ä¹Ÿå¯ä»¥éå†Wordsé›†åˆæ¥è·å–æ–‡æœ¬ [citation:8]
    // QAxObject *words = document->querySubObject("Words");
    // if (words) {
    //     int count = words->dynamicCall("Count()").toInt();
    //     for (int i = 1; i <= count; ++i) {
    //         QAxObject *word = words->querySubObject("Item(int)", i);
    //         if (word) {
    //             allText.append(word->dynamicCall("Text()").toString());
    //             delete word;
    //         }
    //     }
    //     delete words;
    // }

    // 5. å…³é—­æ–‡æ¡£å¹¶é€€å‡ºWord
    document->dynamicCall("Close()");
    wordApp->dynamicCall("Quit()");

    // æ¸…ç†å¯¹è±¡
    delete document;
    delete wordApp;

    return allText;
}
